ghost/core/core/frontend/helpers/get.js

```diff
@@ -131,6 +132,58 @@ function parseOptions(globals, data, options) {
  */
 async function makeAPICall(resource, controllerName, action, apiOptions) {
     const controller = api[controllerName];
+    let makeRequest = () => controller[action](apiOptions);
+
+    // Only do the optimisation on posts
+    if (resource === 'posts' && apiOptions.filter) {
+        try {
+            const parsedFilter = nqlLang.parse(apiOptions.filter);
+            // Support either `id:blah` or `id:blah+other:stuff`
+            if (parsedFilter.$and || parsedFilter.id) {
+                const queries = parsedFilter.$and || [parsedFilter];
+
+                for (const query of queries) {
+                    if ('id' in query) {
+                        if ('$ne' in query.id) {
+                            // This checks that there is only one occurence of a negative id filter
+                            // So we know it's safe to use the `replace` method
+                            if (apiOptions.filter.split('id:-').length === 2) {
+                                const idToFilter = query.id.$ne;
+
+                                // The default limit is 15, the addition order is to cast apiOptions.limit to a number
+                                let limit = apiOptions.limit;
+                                if (apiOptions.limit !== 'all') {
+                                    limit = apiOptions.limit ? 1 + parseInt(apiOptions.limit, 10) : 16;
+                                }
+
+                                // We replace with id:-null so we don't have to deal with leading/trailing AND operators
+                                const filter = apiOptions.filter.replace(/id:-[a-f0-9A-F]{24}/, 'id:-null');
+
+                                makeRequest = async () => {
+                                    const result = await controller[action]({
+                                        ...apiOptions,
+                                        limit,
+                                        filter
+                                    });
+
+                                    const filteredPosts = result?.posts?.filter((post) => {
+                                        return post.id !== idToFilter;
+                                    }) || [];
+
+                                    return {
+                                        ...result,
+                                        posts: limit === 'all' ? filteredPosts : filteredPosts.slice(0, limit - 1)
+                                    };
+                                };
+                            }
+                        }
+                    }
+                }
+            }
+        } catch (err) {
+            logging.warn(err);
```

Comment:
I'm not 100% sure - but the idea is that we log the error and the fall back to the default behaviour - I wanted all of this new logic to be wrapped in the `try/catch` to avoid an unknown errors

Reply:
Oh I actually misread the code, and forgot we're just overriding `makeRequest` here

Reply:
RE: Writing this with NQL tooling - 100% agree, I would love to have a mongo-query -> NQL string helper, that would be soooooo useful, not just here!

I can look at refactoring this to use a separate function - I' do like all of the logic revolving around this being in one place tho

Reply:
@ErisDS Is something like this what you were thinking? https://github.com/TryGhost/Ghost/pull/19865/commits/d53544089b594a3c7ea731b2608c5bc5a10daf47

Reply:
I think the logic needs to be in one file, I don't think it needs to be in one function? This kind of thing benefits from lots of small-fire unit tests IMO. It's hard right now to be certain it does what it says on the tin 😬 

The code to fix the results after is much, much simpler and easier to reason about: 

If this optimisation has been done, then either filter out the right id, or remove the last item.

So the function to do the optimisation could return false if there's no optimisation, and we could use that to determine whether to run the cleanup function or not.

IDK. I'd defo optimise for making the fiddly bit testable.